File: apps/python/design-pattern/src/adapter.py
"""Adapter パターン

BritishTextObject（mm / colour）を TextObject インターフェース
（inches / color）に適合させる。
"""

from typing import Protocol


class TextObject(Protocol):
    """テキストオブジェクトプロトコル（Target）"""

    @property
    def text(self) -> str: ...

    @property
    def size_inches(self) -> float: ...

    @property
    def color(self) -> str: ...


class BritishTextObject:
    """英国式テキストオブジェクト（Adaptee）"""

    def __init__(self, string: str, size_mm: float, colour: str) -> None:
        self.string = string
        self.size_mm = size_mm
        self.colour = colour


class BritishTextObjectAdapter:
    """Adapter: BritishTextObject を TextObject に適合させる"""

    def __init__(self, bto: BritishTextObject) -> None:
        self._bto = bto

    @property
    def text(self) -> str:
        return self._bto.string

    @property
    def size_inches(self) -> float:
        return self._bto.size_mm / 25.4

    @property
    def color(self) -> str:
        return self._bto.colour


class Renderer:
    """レンダラー（Client）"""

    def render(self, text_object: TextObject) -> str:
        return (
            f"text:{text_object.text} "
            f"size:{text_object.size_inches} "
            f"color:{text_object.color}"
        )

File: apps/javascript/design-pattern/src/adapter.js
// Adapter パターン
// 互換性のないインターフェースを変換して接続する

export class TextObject {
  constructor(text, sizeInches, color) {
    this.text = text;
    this.sizeInches = sizeInches;
    this.color = color;
  }
}

export class BritishTextObject {
  constructor(string, sizeMm, colour) {
    this.string = string;
    this.sizeMm = sizeMm;
    this.colour = colour;
  }
}

export class BritishTextObjectAdapter {
  constructor(britishObject) {
    this._britishObject = britishObject;
  }

  get text() {
    return this._britishObject.string;
  }

  set text(value) {
    this._britishObject.string = value;
  }

  get sizeInches() {
    return this._britishObject.sizeMm / 25.4;
  }

  set sizeInches(value) {
    this._britishObject.sizeMm = value * 25.4;
  }

  get color() {
    return this._britishObject.colour;
  }

  set color(value) {
    this._britishObject.colour = value;
  }
}

export class Renderer {
  render(textObject) {
    return `${textObject.text} (${textObject.sizeInches.toFixed(1)}in, ${textObject.color})`;
  }
}
